virtio block: Implement the block backend driver for virtio

luodeb opened the pull request:
This crate provides a VirtIO block device implementation for the AxVirtIO framework.
It includes MMIO transport, block device backend traits, and request handling for VirtIO block devices according to the VirtIO specification.

luodeb committed:
feat(block): improve block device backend and request handling

- Update file backend to use separate read and write file handles
- Implement safe guest memory access using axvirtio_common::memory
- Add logging and error handling in block request processing
- Refactor VirtioMmioDevice for better queue handling and error reporting
- Update constants and configuration for block device

luodeb committed:
build: remove unused workspace members

- Remove 'axvirtio-net' and 'axvirtio-console' from workspace members
- This change simplifies the project structure by eliminating unused components

luodeb committed:
build: remove unused dependencies from Cargo.toml

- Remove axvirtio-net and axvirtio-console dependencies
- Keep axvirtio-common and axvirtio-blk dependencies

luodeb commented:
TODO：Decouple the parts dependent on `axvisor_api` and `axstd` by implementing them with generics.  
The backend implementation should not be placed in this part.

luodeb committed:
chore(workspace): update homepage URL for axvirtio

- Set homepage to "https://github.com/arceos-hypervisor/axvirtio" for the axvirtio workspace
- This change improves discoverability and provides a direct link to the project's GitHub repository

luodeb committed:
refactor(axvirtio): remove unused dependencies and API calls

- Remove unused dependency on axhal
- Remove path dependency on axvisor_api
- Delete unused legacy functions in axvirtio-common
- Comment out translate_guest_to_host function in GuestMemoryAccess implementation

luodeb committed:
refactor(axvirtio-blk): rewrite block device implementation

- Remove file and memory backends
- Inject block backend and guest memory access as type parameters
- Update VirtioMmioDevice to use new backend and memory access traits
- Simplify block request processing and error handling
- Remove unused code and simplify existing code

luodeb committed:
docs(axvirtio-common): improve crate documentation and clean up code comments

- Add comprehensive crate-level documentation
- Reorganize and clarify module-level comments

luodeb committed:
docs(axvirtio-blk): add crate-level documentation

- Add comprehensive documentation for the AxVirtIO Block Device Library
- Include features, usage examples, and trait implementations
- Improve crate usability and developer understanding

luodeb committed:
refactor(memory): replace GuestMemoryAccess with AddressTranslator

- Update BlockRequest to use AddressTranslator instead of GuestMemoryAccess
- Modify VirtioMmioDevice to use AddressTranslator for memory operations
- Update queue structures (AvailableRing, DescriptorTable, VirtioQueue, UsedRing) to use AddressTranslator
- Introduce GuestMemoryAccessor to encapsulate address translation and memory access

luodeb committed:
docs(axvirtio-blk): update documentation and rename parameter

- Update usage examples in axvirtio-blk documentation
- Add documentation for new parameters and traits
- Rename 'memory' parameter to 'transtor' in VirtioMmioDevice::new

aarkegz commented on axvirtio-common/src/memory.rs:
```diff
@@ -0,0 +1,113 @@
+use memory_addr::PhysAddr;
+
+/// Trait for address translation
+pub trait AddressTranslator {
```
@hky1999 Maybe we should extract this to `axvisor_api` or `axaddrspace`, depending on in what form should it exist?

luodeb replied:
I think it's best not to rely on any specific components of `axvisor/arceos`, so that in the future it can be used not only in axvisor but also in any other hypervisor.

hky1999 replied:
> I think it's best not to rely on any specific components of `axvisor/arceos`, so that in the future it can be used not only in axvisor but also in any other hypervisor.

I partialy agree with @aarkegz on extract this to `axaddrspace`, because it's not designed to be a special component of `axvisor/arceos`, but a general crate design to cover all virtual address space management, including both normal stage-1 address translation for user process and virtualized stage-2 address translation for guest VM.

As a matter of fact, the `AddressTranslator` and `xxx_accessor` generic type should be general to user access and "guest VM" access

aarkegz commented on axvirtio-common/src/memory.rs:
```diff
@@ -0,0 +1,113 @@
+/// Trait for safe guest memory access operations
+pub trait GuestMemoryAccess {
+    /// Translate a guest physical address to host physical address
+    fn translate_guest_to_host(&self, guest_addr: GuestPhysAddr) -> Option<PhysAddr>;
```
Is is necessary to specify the VM ID here?

luodeb replied:
Not needed. It's implemented by `axvisor_api`, where the specific VM is specified during the instantiation of virtio-blk.

aarkegz replied:
I think a stateful MemoryAccessor is bad. How's your opinion? @hky1999

aarkegz commented on axvirtio-common/src/memory.rs:
```diff
@@ -0,0 +1,113 @@
+
+/// Guest memory access with injected translator
+#[derive(Debug, Clone)]
+pub struct GuestMemoryAccessor<T> {
```
This design is somehow weird. Is it possible to `impl<T: AddressTranslator> GuestMemoryAccessor for T`?

hky1999 replied:
@luodeb

aarkegz commented on axvirtio-common/src/memory.rs:
```diff
@@ -0,0 +1,113 @@
+        self.translator.translate_guest_to_host(guest_addr)
+    }
+
+    fn read_obj<U: Copy>(&self, guest_addr: GuestPhysAddr) -> VirtioResult<U> {
```
It lacks cross-page-boundary handling capability.

luodeb replied:
How do I fix this?

aarkegz commented on axvirtio-common/src/device_type.rs:
```diff
@@ -0,0 +1,68 @@
+/// VirtIO device types (simplified version)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[repr(u32)]
+pub enum VirtioDeviceType {
```
Maybe just call this type `VirtioDeviceID`?

aarkegz replied:
To keep consistency with the VirtIO Spec?

luodeb replied:
ok

aarkegz commented on axvirtio-common/src/device_type.rs:
```diff
@@ -0,0 +1,68 @@
+
+impl VirtioDeviceType {
+    /// Convert device ID to device type
+    pub fn from_device_id(device_id: u32) -> Self {
```
And call this fn `from_u32/usize`?

luodeb replied:
Maybe `impl From<u32>` is better

aarkegz commented on axvirtio-common/src/error.rs:
```diff
@@ -37,6 +37,10 @@ pub enum VirtioError {
+    NotFound,
 }
 
 impl VirtioError {
```
Also `impl From<VirtioError> for AxError`?

luodeb replied:
Is it necessary for this crate to depend on `AxError`? May this be implemented in `axdevice` instead?

aarkegz replied:
But you have already provided a custom conversion function below

aarkegz commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -45,24 +47,34 @@ pub enum DataSource {
     /// Data source (buffer or guest memory)
     pub data_source: DataSource,
+    /// Guest memory accessor
+    memory: GuestMemoryAccessor<T>,
```
See my comment on these two traits. The design is weird here. Is it necessary (at least beneficial) to make AddressTranslator stateful?

aarkegz replied:
And do we really need to inject different `AddressTranslator`s to different VirtIO devices?

aarkegz replied:
If really so (though I personally doubt), use `dyn Trait`.

luodeb replied:
The main purpose is still to avoid depending on `axvisor_api`, which is why an alternative approach is needed to obtain the address translation function.

aarkegz replied:
I don't mean the place for this Trait. I mean having a Accessor field here is somehow weird, does it really worth it? Do we really need more than one different accessors?

luodeb replied:
It's my fault. It really shouldn't be.

aarkegz commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -45,24 +47,34 @@ pub enum DataSource {
 
+/// Block request processing result
+#[derive(Debug)]
+pub struct BlockRequestResult {
```
Is this wrapper worth it? Can we use `pub type BlockRequestResult = u8` or does we need to add more fields into it in the future?

aarkegz replied:
Or can we make it an `enum`?

luodeb replied:
emmm, maybe `enum` is better

aarkegz commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -106,15 +118,10 @@ impl BlockRequest {
         buffers: &[(GuestPhysAddr, usize, bool)],
         _status_addr: GuestPhysAddr,
     ) -> u8 {
```
Change the result type to `BlockRequestResult`. Same for `handle_read/write/flush_request_guest_memory`.

luodeb replied:
ok

aarkegz commented on axvirtio-common/src/mmio/transport.rs:
```diff
@@ -1,7 +1,7 @@
 use axerrno::AxResult;
 
-use crate::error::VirtioError;
+use crate::{error::VirtioError, VIRTIO_MMIO_CONFIG};
```
What is `VIRTIO_MMIO_CONFIG`? This const needs a better name.

luodeb replied:
`VIRTIO_MMIO_CONFIG_OFFSET` is better

aarkegz commented on axvirtio-common/src/mmio/transport.rs:
```diff
@@ -92,14 +88,10 @@ impl MmioTransport {
-                    data[0], data[1], data[2], data[3], data[4], data[5], data[6], data[7],
-                ]) as usize
-            }
+            AccessWidth::Dword => u32::from_le_bytes([data[0], data[1], data[2], data[3]]) as usize,
```
Why not 

```
let (data, _) = data.split_first_chunk().ok_or_else(|| SomeError)?;
return Ok(u32::from_le_bytes(*data));
```

luodeb replied:
ok

aarkegz commented on axvirtio-common/src/queue/available.rs:
```diff
@@ -1,20 +1,16 @@
 #[repr(C)]
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Clone, Copy, Default)]
 pub struct VirtqAvail {
```
Can we name it `VirtQueueAvail`?

luodeb replied:
ok

aarkegz commented on axvirtio-common/src/queue/mod.rs:
```diff
@@ -4,60 +4,100 @@ mod used;
+        addr: GuestPhysAddr,
+        memory: &impl GuestMemoryAccess,
+    ) -> VirtioResult<Self> {
+        let request_type: u32 = memory.read_obj(addr)?;
```
Can we just read a `#[repr(C)] VirtioBlockHeader` from guest memory

luodeb replied:
ok

luodeb committed:
refactor(axvirtio): rewrite error handling and device type

- Remove axerrno dependency
- Rename VirtioDeviceType to VirtioDeviceID
- Update error handling to use custom VirtioResult type
- Adjust device configuration and queue structures
- Modify block request processing and status handling

luodeb committed:
refactor(memory): generalize read_obj function and improve VirtioBlockHeader

- Generalize the read_obj function in GuestMemoryAccessor to work with any Copy type
- Add repr(C) attribute to VirtioBlockHeader for consistent memory layout
- Simplify VirtioBlockHeader reading logic in virtio_block_request function

luodeb committed:
refactor(axvirtio): use Arc<GuestMemoryAccessor> instead of GuestMemoryAccessor

- Replace GuestMemoryAccessor with Arc<GuestMemoryAccessor> in multiple files
- Update related imports and usages
- Rename VirtioMmioDevice to VirtioMmioBlockDevice
- Adjust function signatures and implementations to use the new Arc type

luodeb committed:
docs(axvirtio-blk): update device struct name in usage example

- Rename `VirtioMmioDevice` to `VirtioMmioBlockDevice` in the usage example
- This change aligns the example with the actual struct name used in the library

luodeb committed:
refactor(memory): remove GuestMemoryAccess trait and simplify memory access

- Remove GuestMemoryAccess trait from memory module
- Simplify GuestMemoryAccessor implementation
- Update related modules to use new memory access methods
- Improve code readability and reduce redundancy

hky1999 commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -1,7 +1,12 @@
+    memory::{AddressTranslator, GuestMemoryAccessor},
+    VirtioResult,
+};
+use log::{debug, error, trace, warn};
```
in lib.rs in this crate
use
```rust
#[macro_use]
extern crate log;
```

hky1999 commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -12,6 +17,8 @@ pub enum BlockRequestType {
     /// Flush request
     Flush,
+    /// Unsupported request
+    Unsupported,
```
Implement it as default

use `Impl Default for BlockRequestType`

hky1999 commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -82,19 +105,23 @@ impl BlockRequest {
+                // Write status to guest memory using injected memory accessor
+                if self
+                    .accessor
+                    .write_obj(*status_addr, status.clone() as u8)
```
what is the type of `status`,  can we remove the `as u8`

hky1999 commented on axvirtio-common/src/config.rs:
```diff
@@ -20,88 +19,48 @@ pub struct VirtioConfig {
 
     /// Create a new block device configuration
-    pub fn new_block_device(device_index: usize) -> Self {
+    pub fn new_block_device(base_ipa: usize) -> Self {
```
replace the type of `base_ipa` (or similar vars) from `usize` to `GuestPhysAddr` from `axaddrspace` to avoid confusion.

hky1999 commented on axvirtio-common/src/queue/mod.rs:
```diff
@@ -74,7 +110,7 @@ impl VirtioQueue {
     pub fn set_desc_table_addr(&mut self, addr: GuestPhysAddr) -> VirtioResult<()> {
         self.desc_table_addr = addr;
         if addr.as_usize() != 0 {
```
return Error if `addr` has been set before.

hky1999 commented on axvirtio-common/src/queue/available.rs:
```diff
@@ -40,22 +37,25 @@ impl VirtqAvail {
 /// Available ring management
 #[derive(Debug, Clone)]
-pub struct AvailableRing {
+pub struct AvailableRing<T: AddressTranslator + Clone> {
```
It is so strange to use `T: AddressTranslator`

luodeb committed:
refactor(axvirtio): update crate structure and improve code quality

- Update Cargo.toml files across multiple crates
- Refactor axvirtio-blk and axvirtio-common crates
- Improve error handling and logging
- Optimize data structures and function signatures
- Remove redundant code and improve code readability

luodeb committed:
unit test fix

luodeb committed:
refactor(axvirtio): improve code structure and readability for virtio-blk device

- Rename and reorganize function parameters for better clarity
- Move request header parsing to the top of the request processing function
- Remove redundant device_id field from VirtioConfig
- Update MIN_DESCRIPTOR_CHAIN_LENGTH to reflect actual minimum length

luodeb committed:
fix(virtio-blk): improve status handling and feature negotiation

- Remove unnecessary status write in BlockRequest::execute
- Set device features in VirtioMmioBlockDevice::new
- Fix error status handling in request processing
- Add status descriptor length check in DescriptorTable::get_status_addr
- Implement feature negotiation validation in VirtioMmioBlockDevice::update_status

luodeb committed:
feat(axvirtio-blk): add support for configuring VirtIO block device features

- Introduce VirtioBlockConfig struct to allow users to specify device features
- Update VirtioMmioBlockDevice::new function to accept VirtioBlockConfig
- Modify driver_features handling to check against device_features

aarkegz commented on axvirtio-common/src/config.rs:
```diff
@@ -20,88 +17,40 @@ pub struct VirtioConfig {
-    /// Get the console device path for console devices
-    pub fn get_console_path(&self) -> alloc::string::String {
-        self.get_device_path("console", "sock")
+        Self::new(base_ipa, features, 1, VirtioDeviceID::Block)
```
Add a const for `1`.

aarkegz commented on axvirtio-common/src/config.rs:
```diff
@@ -20,88 +17,40 @@ pub struct VirtioConfig {
 impl Default for VirtioConfig {
     fn default() -> Self {
-        Self::new_block_device(0)
+        Self::new_block_device(GuestPhysAddr::from(VIRTIO_MMIO_BASE))
```
Need more explanation for `pub const VIRTIO_MMIO_BASE: usize = 0x0a00_0000;` and why it should be used as the default value.

aarkegz commented on axvirtio-common/src/config.rs:
```diff
@@ -20,88 +17,40 @@ pub struct VirtioConfig {
     }
 }
 
 impl Default for VirtioConfig {
```
Implementing `Default` for `VirtioConfig` seems to be a little bit strange.

luodeb committed:
refactor(memory): remove GuestMemoryAccessor and use AddressTranslator directly

- Remove GuestMemoryAccessor struct and its implementations
- Use AddressTranslator directly in BlockRequest, VirtioMmioBlockDevice, and queue modules
- Update related functions and methods to work with AddressTranslator
- Simplify memory access operations by removing indirection through GuestMemoryAccessor

luodeb committed:
refactor(axvirtio-blk): move parse_virtio_block_header to device.rs

- Move parse_virtio_block_header function from queue/mod.rs to device.rs
- Update related imports and usages
- This change prepares for implementing header parsing in the transport layer

luodeb committed:
refactor(axvirtio-blk): remove unused constant re-exports

- Remove re-export of constants for backward compatibility
- This change simplifies the code and removes unnecessary exports

luodeb committed:
refactor(axvirtio-common): remove unused VirtioBlockHeader structure

- Removed VirtioBlockHeader definition from the queue module
- This structure was likely unused and has been removed to simplify the codebase

luodeb committed:
refactor(axvirtio): Replace AddressTranslator with GuestMemoryAccessor

aarkegz commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -20,7 +24,7 @@ impl From<u32> for BlockRequestType {
             VIRTIO_BLK_T_OUT => BlockRequestType::Write,
             VIRTIO_BLK_T_FLUSH => BlockRequestType::Flush,
-            _ => BlockRequestType::Read, // Default to read for unknown types
+            _ => BlockRequestType::Unsupported, // Default to read for unknown types
```
The comment is wrong here.

aarkegz commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -31,6 +35,7 @@ impl From<BlockRequestType> for u32 {
             BlockRequestType::Read => VIRTIO_BLK_T_IN,
             BlockRequestType::Write => VIRTIO_BLK_T_OUT,
             BlockRequestType::Flush => VIRTIO_BLK_T_FLUSH,
+            BlockRequestType::Unsupported => VIRTIO_BLK_T_IN,
```
Please consider it carefully.

aarkegz commented on axvirtio-blk/src/block/request.rs:
```diff
@@ -176,31 +187,31 @@ impl BlockRequest {
-                log::warn!("Data buffer exceeds write data range");
-                return VIRTIO_BLK_S_IOERR;
+                warn!("Data buffer exceeds write data range");
+                return Ok(BlockRequestResult::Unsupported);
```
Shouldn't it be an error?

aarkegz commented on axvirtio-blk/src/mmio/device.rs:
```diff
@@ -1,19 +1,17 @@
+use crate::{backend::BlockBackend, mmio::VirtioBlockHeader};
 use axvirtio_common::{MmioTransport, VirtioConfig, VirtioError, VirtioQueue, VirtioResult};
 
 /// VirtIO MMIO device state
```
Bad comment.

aarkegz commented on axvirtio-common/src/mmio/transport.rs:
```diff
@@ -1,17 +1,16 @@
+use crate::{error::VirtioError, VirtioResult, VIRTIO_MMIO_CONFIG_OFFSET};
 
 /// MMIO transport layer utilities
 pub struct MmioTransport;
```
Maybe it's better to make this struct a mod?

aarkegz commented on axvirtio-common/src/mmio/transport.rs:
```diff
@@ -88,17 +83,31 @@ impl MmioTransport {
-            AccessWidth::Byte => data[0] as usize,
-            AccessWidth::Word => u16::from_le_bytes([data[0], data[1]]) as usize,
+            AccessWidth::Byte => {
+                let (bytes, _) = data
```
Is it possible to introduce a macro for these similar codes?

aarkegz commented on axvirtio-common/src/queue/available.rs:
```diff
@@ -40,22 +37,25 @@ impl VirtqAvail {
 /// Available ring management
 #[derive(Debug, Clone)]
-pub struct AvailableRing {
+pub struct AvailableRing<T: GuestMemoryAccessor + Clone> {
```
Add a comment to explain the relationship between VirtQueueAvail and AvailableRing

aarkegz replied:
Same for used.

luodeb committed:
refactor(axvirtio): improve error handling and documentation for block device operations

luodeb committed:
code format

luodeb committed:
chore: update edition to 2024 in Cargo.toml

luodeb committed:
refactor: reorder use statements for consistency and clarity

luodeb committed:
docs: fix code block formatting in documentation

luodeb committed:
docs: fix formatting of array indices in VirtIO ring documentation

luodeb committed:
chore: update axaddrspace version to 0.2.0 and add Default trait to VirtioDeviceID

luodeb committed:
feat: add README and integration tests for axvirtio-blk
